mdformat_pyproject/plugin.py

```diff
@@ -23,63 +24,94 @@
+
+    Search the first ``.mdformat.toml`` or ``pyproject.toml`` file in the provided
+    ``search_path`` folder or its parents.
+
+    The search is done ascending through the folders tree until a ``.mdformat.toml`` or a
+    ``pyproject.toml`` file is found. If the root ``/`` is reached, ``None`` is returned.
 
-    Notice that if `.mdformat.toml` exists it is ignored.
+    ``.mdformat.toml`` takes precedence over ``pyproject.toml`` if both are present in the
+    same folder. In that case, a warning is logged to inform the user that the
+    ``pyproject.toml`` file has been ignored.
+
+    ``pyproject.toml`` files without a ``[tool.mdformat]`` section are ignored.
+
+    This behavior mimics the one from Ruff, as described in `issues#17
+    <https://github.com/csala/mdformat-pyproject/issues/17>`_.
     """
-    pyproject_path = _find_pyproject_toml_path(conf_dir)
-    if pyproject_path:
-        pyproject_opts = _parse_pyproject(pyproject_path)
-    else:
-        pyproject_opts = {}
+    if search_path.is_file():
+        search_path = search_path.parent
 
-    return pyproject_opts, pyproject_path
+    for parent in (search_path, *search_path.parents):
+        # Try to find a pyproject.toml file first, with a [tool.mdformat] section.
+        pyproject_options = None
+        pyproject_path = parent / "pyproject.toml"
+        if pyproject_path.is_file():
+            pyproject_options = _parse_pyproject(pyproject_path)
+
+        # Read options from .mdformat.toml if present.
+        mdformat_path = parent / ".mdformat.toml"
+        if mdformat_path.is_file():
+            # If both pyproject.toml and .mdformat.toml are present, the latter takes
+            # precedence, but we warn the user that pyproject.toml is being ignored.
+            if pyproject_options is not None:
+                print_paragraphs(
+                    (
+                        f"Warning: ignoring mdformat options from {pyproject_path} "
+                        f"since {mdformat_path} is present.",
+                    )
+                )
+
+            # Return options from .mdformat.toml using the original function, even if
+            # they are empty.
+            return _orig_read_toml_opts(mdformat_path.parent)
+
+        # Return options from pyproject.toml if .mdformat.toml is not present.
+        elif pyproject_options is not None:
+            return pyproject_options, pyproject_path
+
+    # No config file found, return empty options.
+    return {}, None
 
 
 def update_mdit(mdit: markdown_it.MarkdownIt) -> None:
-    """No-op, since this plugin only monkey patches and does not modify mdit."""
+    """No-op, since this plugin only monkey patches and does not modify ``mdit``."""
     pass
 
 
 RENDERERS: dict[str, Render] = {}
 
 # Monkey patch mdformat._conf to use our own read_toml_opts version
-mdformat._conf.read_toml_opts = read_toml_opts
+mdformat._cli.read_toml_opts = patched_read_toml_opts
```

Comment:
This is critical! It turns out that we were patching the wrong `read_toml_opts`, and the current plugin version is actually not reading the options correctly!